src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1015,3 +1069,919 @@ def upgrade_agents(cmd, client, resource_group_name, cluster_name, kube_config=N
+        try:
+            os.remove(config_file_location)
+        except Exception as e:
+            telemetry.set_user_fault()
+            telemetry.set_exception(exception=e, fault_type=consts.Remove_Config_Fault_Type,
+                                    summary='Unable to remove old config file')
+            raise CLIError("Failed to remove old config." + str(e))
+
+    # initializations
+    user_type = 'sat'
+    creds = ''
+
+    # if service account token is not passed
+    if token is None:
+        # Identifying type of logged in entity
+        account = get_subscription_id(cmd.cli_ctx)
+        account = Profile().get_subscription(account)
+        user_type = account['user']['type']
+
+        tenantId = _graph_client_factory(cmd.cli_ctx).config.tenant_id
+
+        if user_type == 'user':
+            dict_file = {'server': {'httpPort': int(client_proxy_port), 'httpsPort': int(api_server_port)}, 'identity': {'tenantID': tenantId, 'clientID': consts.CLIENTPROXY_CLIENT_ID}}
+        else:
+            dict_file = {'server': {'httpPort': int(client_proxy_port), 'httpsPort': int(api_server_port)}, 'identity': {'tenantID': tenantId, 'clientID': account['user']['name']}}
+
+        if cloud == 'DOGFOOD':
+            dict_file['cloud'] = 'AzureDogFood'
+
+        # Fetching creds
+        creds_location = os.path.expanduser(os.path.join('~', creds_string))
+        try:
+            with open(creds_location) as f:
+                creds_list = json.load(f)
+        except Exception as e:
+            telemetry.set_user_fault()
+            telemetry.set_exception(exception=e, fault_type=consts.Load_Creds_Fault_Type,
+                                    summary='Unable to load accessToken.json')
+            raise CLIError("Failed to load credentials." + str(e))
+
+        user_name = account['user']['name']
+        loop_flag = False
+        for i in range(len(creds_list)):
+            if loop_flag:
+                break
+            creds_obj = creds_list[i]
+            for key in creds_obj:
+                if user_type == 'user':
+                    if key == 'refreshToken':
+                        creds = creds_obj[key]
+                    elif key == 'userId':
+                        if creds_obj[key] == user_name:
+                            loop_flag = True
+                else:
+                    if key == 'accessToken':
+                        creds = creds_obj[key]
+                    elif key == 'servicePrincipalId':
+                        if creds_obj[key] == user_name:
+                            loop_flag = True
+
+        if creds == '':
+            telemetry.set_user_fault()
+            telemetry.set_exception(exception='Credentials of user not found.', fault_type=consts.Creds_NotFound_Fault_Type,
+                                    summary='Unable to find creds of user')
+            raise CLIError("Credentials of user not found.")
+
+        if user_type != 'user':
```

Comment:
so will we have only one spn in the accessToken.json ?.

Reply:
No, we can have multiple spn's. Thats why we are running a loop above and selecting only that spn whose username matches with the logged in spn.